### PR TITLE
Logic

### PR DESCRIPTION
**Pull request by arerererere**
basic classes logic game

### CODE REVIEW

File: src/classes/Bullet.ts [hunk 1]
```diff
@@ -15,82 +16,83 @@ export class Bullet {
...
+
+  public constructor(owner: Tank, id: number) {
     this.id = id;
     this.owner = owner;
     this.direction = owner.direction;
-    while(this.direction > 360) this.direction -= 360;
-    while(this.direction < 0) this.direction += 360;
+    while (this.direction > 360) {
+      this.direction -= 360;
+    }
```
Comment (MarozauArtsiom) [inline src/classes/Bullet.ts]: this.direction = this.direction % 360

Comment (MarozauArtsiom) [reply src/classes/Bullet.ts]: replace 360 with const

File: src/classes/Bullet.ts [hunk 2]
```diff
@@ -15,82 +16,83 @@ export class Bullet {
...
+    while (this.direction < 0) {
+      this.direction += 360;
```
Comment (MarozauArtsiom) [inline src/classes/Bullet.ts]: if (this.direction < 0) {
    this.direction = 360 - (math.abs(this.direction) % 360)
}

File: src/classes/CollisionSolution.ts [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+      }
+    }
+
+  }
+
+  public scanBullets(tank: Tank): void {
+    const dirThis: number = (tank.direction) * Math.PI / 180;
+    const coefA: number = CONFIG.smallVisionSide;
+    const coefB: number = CONFIG.bigVisionSide;
+    const xA: number = Math.round(coefA * Math.sin(dirThis) - coefA * Math.cos(dirThis) + tank.x);
```
Comment (MarozauArtsiom) [inline src/classes/CollisionSolution.ts]: looks like this method is duplication of scanWalls. If so, then make a separate method

File: src/classes/IdentificatorAi.ts [hunk 4]
```diff
@@ -64,6 +62,7 @@ export class IdentificatorAi {
 
 export function getRandomId(): number {
   let idRandomString: string = (new Date()).getTime().toString();
-  idRandomString = idRandomString.substr(idRandomString.length - 6, 6) + "" + Math.round(10000000 * Math.random());
+  idRandomString = idRandomString.substr(idRandomString.length - 6, 6) + '' + Math.round(10000000 * Math.random());
```
Comment (MarozauArtsiom) [inline src/classes/IdentificatorAi.ts]: too many numeric literals, use const instead

File: src/classes/CollisionSolution.ts [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+
+  public hitEnemyTestBullet(bullet: Bullet): boolean {
+    let hitEnemyTest: boolean = false;
+    for (const item of this.tankList) {
+      if (item.x === bullet.x && item.y === bullet.y) {
+        if (item.id !== bullet.owner.id) {
+          hitEnemyTest = true;
+          item.health = item.health - bullet.power;
+          if (item.health <= 0) {
+            bullet.owner.killsScore();
```
Comment (MarozauArtsiom) [inline src/classes/CollisionSolution.ts]: what does it mean? what this method must do? may be you need to think about better name

File: src/classes/CollisionSolution.ts [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+  public constructor(battlefield: Battlefield) {
+    this.wallList = [0, 0, 0, 0];
+    this.tankList = [];
+    this.bulletList = [];
+    this.battlefield = battlefield;
+  }
+
+  public updateBattlefield(battlefield: Battlefield): void {
+    this.battlefield = battlefield;
+    const wall: number[] = [];
```
Comment (MarozauArtsiom) [inline src/classes/CollisionSolution.ts]: I see that you initialize array but it never used

File: src/classes/Bullet.ts [hunk 7]
```diff
@@ -15,82 +16,83 @@ export class Bullet {
...
-    this.x += Math.round(this.speed*Math.cos(this.direction*(Math.PI/180)));
-    this.y += Math.round(this.speed*Math.sin(this.direction*(Math.PI/180)));
+  public moveForward(): void {
+    this.x += Math.round(Math.cos(this.direction * (Math.PI / 180)));
+    this.y += Math.round(Math.sin(this.direction * (Math.PI / 180)));
   }
 
-  onDestroy(): void {
+  public onDestroy(): void {
     this.owner.bullets.filter((b) => {
```
Comment (MarozauArtsiom) [inline src/classes/Bullet.ts]: I dont see any purpose for this filter

Comment (MarozauArtsiom) [inline src/classes/Bullet.ts]: Cant understand why you make this filter. Filter return new instance of array, it doesnt modify existing one

File: src/classes/CollisionSolution.ts (hunk 6 above)
Comment (arerererere) [reply src/classes/CollisionSolution.ts]: deleted. this method was planned to test the reduction of the zone

File: src/classes/CollisionSolution.ts (hunk 5 above)
Comment (arerererere) [reply src/classes/CollisionSolution.ts]: ok. changed to addKillsScore()

File: src/classes/CollisionSolution.ts (hunk 3 above)
Comment (arerererere) [reply src/classes/CollisionSolution.ts]: No. these methods simply use one algorithm to determine the zone of visibility. Has deduced algorithm of definition of a zone of visibility in a separate method.

File: src/classes/Bullet.ts (hunk 1 above)
Comment (arerererere) [reply src/classes/Bullet.ts]: originally planned to check the direction. I will delete this code. It is not necessary.

File: src/classes/Bullet.ts (hunk 2 above)
Comment (arerererere) [reply src/classes/Bullet.ts]: originally planned to check the direction. I will delete this code. It is not necessary. 
Yes, I had to improve) like this.
if (this.direction < 0 ||  this.direction > 360) {
   this.direction = 0 + (math.abs(this.direction) % 360)
} 
Thanks for your comment.

File: src/classes/Bullet.ts (hunk 7 above)
Comment (arerererere) [reply src/classes/Bullet.ts]: sorry. this is not special)

File: src/main.ts [hunk 8]
```diff
@@ -0,0 +1,7 @@
+import { Greeter } from './greeter';
```
Comment (vharadkou) [inline src/main.ts]: for what reason this file was restored.

File: src/classes/Battlefield.ts [hunk 9]
```diff
@@ -25,16 +20,16 @@ export class Battlefield {
     this.finishY = this.startY + height - 1;
   }
 
-  setSize(width: number, height: number): void {
+  public setSize(width: number, height: number): void {
```
Comment (YegorFeoktistov) [inline src/classes/Battlefield.ts:23]: In this method you should recalculate finishX and finishY values.

**Review by vharadkou — APPROVED** [review]

**Review by MarozauArtsiom — APPROVED** [review]

**Pull request merged**